nistats/reporting/glm_reporter.py

```diff
@@ -0,0 +1,783 @@
+            contrasts,
+            title,
+            )
+    with pd.option_context('display.max_colwidth', 100):
+        model_attributes_html = _make_attributes_table(model)
+    statistical_maps = make_stat_maps(model, contrasts)
+    html_design_matrices = _make_svg_url_design_matrices(design_matrices)
+    roi_plot_html_code = _make_roi_svg(roi_img, bg_img)
+    all_components = _make_report_components(
+            stat_img=statistical_maps,
+            contrasts_plots=contrast_plots,
+            threshold=threshold,
+            alpha=alpha,
+            cluster_threshold=cluster_threshold,
+            height_control=height_control,
+            min_distance=min_distance,
+            bg_img=bg_img,
+            display_mode=display_mode,
+            plot_type=plot_type,
+            )
+    all_components_text = '\n'.join(all_components)
+    report_values = {'page_title': escape(page_title),
+                     'page_heading_1': page_heading_1,
+                     'page_heading_2': page_heading_2,
+                     'model_attributes': model_attributes_html,
+                     'all_contrasts_with_plots': ''.join(
+                         contrast_plots.values()),
+                     'design_matrices': html_design_matrices,
+                     'roi_plot': roi_plot_html_code,
+                     'component': all_components_text,
+                     }
+    report_text = report_template.safe_substitute(**report_values)
+    report = HTMLDocument(report_text)
+    report.width = nb_width  # better visual experience in Jupyter Notebooks.
+    report.height = nb_height
+    return report
+
+
+def _make_contrasts_dict(contrasts):
+    """ Accepts contrasts and returns a dict of them.
+    with the names of contrasts as keys.
+    
+    If contrasts is:
+      dict then returns it unchanged.
+      
+      string or collection of Strings or Sequence[int],
+      returns a dict where key==values
+    
+    Parameters
+    ----------
+    contrasts: String or Collection[str or Sequence[Int]] or Dict[str, str or np.array]
+        Contrast information. Can be a dict of the form
+         {'contrast_name_1': contrast list/array1, 'contrast_name_1': contrast list/array1}
+         ['contast_name_1', 'contast_name_2', ...]
+         [contrast_array_1, contrast_array_2, ...]
+         'contrast_name'
+    
+    Returns
+    -------
+    contrasts: Dict[str, np.array or str]
+        Contrast information, as a dict in the form
+            {'contrast_title_1': contrast_info_1/title_1, ...}
+    """
+    if not isinstance(contrasts, dict):
+        excluded_types = (list, tuple, np.ndarray, str)
+        if isinstance(contrasts, excluded_types):
+            if not isinstance(contrasts[0], excluded_types):
+                contrasts = [contrasts]
+        contrasts = [contrasts] if isinstance(contrasts, str) else contrasts
+        contrasts = {str(contrast_): contrast_ for contrast_ in contrasts}
```

Comment:
This function is fairly general and has nothing to do with contrasts in its generality: should be something like `_coerce_to_dict` ?

Reply:
Oooh excellent name! I'll use that.